File: rerun-reports/lane-a/032/20260828-033651/results/storage/data-plane/java/streaming/storage-dp-java-blob-event-notifier/java-azure-skills/azure-skill-mcp-microsoft-skill/generated-code/src/main/java/com/example/blobevents/receiver/EventReceiver.java
package com.example.blobevents.receiver;

import com.example.blobevents.blob.BlobEventHandler;
import com.example.blobevents.model.BlobLifecycleEvent;

import java.util.List;
import java.util.logging.Logger;

public final class EventReceiver {
    private static final Logger LOGGER = Logger.getLogger(EventReceiver.class.getName());
    private static final String BLOB_CREATED = "Microsoft.Storage.BlobCreated";
    private static final String BLOB_DELETED = "Microsoft.Storage.BlobDeleted";

    private final BlobEventHandler handler;

    public EventReceiver(BlobEventHandler handler) {
        this.handler = handler;
    }

    public List<BlobLifecycleEvent> receive(String jsonPayload) {
        List<BlobLifecycleEvent> events = EventPayloadParser.parse(jsonPayload);
        events.forEach(this::route);
        return events;
    }

    private void route(BlobLifecycleEvent event) {
        switch (event.type()) {
            case BLOB_CREATED -> handler.handleCreated(event);
            case BLOB_DELETED -> handler.handleDeleted(event);
            default -> LOGGER.warning(() -> "Ignoring unrecognized event type: " + event.type());
        }
    }
}

File: rerun-reports/lane-b/072/20260828-065704/results/storage/data-plane/python/streaming/storage-dp-python-blob-event-notifier/python-azure-skills/azure-skill-mcp-microsoft-skill/generated-code/blob_event_notifier/event_receiver.py
from __future__ import annotations

import json
import logging
from collections.abc import Mapping, Sequence
from typing import Any, Protocol

from azure.core.messaging import CloudEvent
from azure.eventgrid import EventGridEvent

from .blob_event_handler import StructuredEvent

BLOB_CREATED = "Microsoft.Storage.BlobCreated"
BLOB_DELETED = "Microsoft.Storage.BlobDeleted"

JsonPayload = str | bytes | bytearray | Mapping[str, Any] | Sequence[Mapping[str, Any]]


class EventHandler(Protocol):
    def handle_created(self, event: StructuredEvent) -> None: ...

    def handle_deleted(self, event: StructuredEvent) -> None: ...


class AsyncEventHandler(Protocol):
    async def handle_created(self, event: StructuredEvent) -> None: ...

    async def handle_deleted(self, event: StructuredEvent) -> None: ...


def deserialize_events(payload: JsonPayload) -> list[StructuredEvent]:
    # JSON decoding only unwraps the webhook batch; SDK helpers construct and validate each event.
    decoded: Any = json.loads(payload) if isinstance(payload, (str, bytes, bytearray)) else payload
    items = [decoded] if isinstance(decoded, Mapping) else list(decoded)

    events: list[StructuredEvent] = []
    for item in items:
        if not isinstance(item, Mapping):
            raise ValueError("Each webhook batch item must be a JSON object")
        if item.get("specversion") == "1.0":
            events.append(CloudEvent.from_dict(dict(item)))
        elif "eventType" in item:
            events.append(EventGridEvent.from_dict(dict(item)))
        else:
            raise ValueError("Payload item is neither CloudEvents 1.0 nor Event Grid schema")
    return events


def _event_type(event: StructuredEvent) -> str:
    return event.type if isinstance(event, CloudEvent) else event.event_type


def receive_events(
    payload: JsonPayload,
    handler: EventHandler,
    logger: logging.Logger | None = None,
) -> list[StructuredEvent]:
    log = logger or logging.getLogger(__name__)
    events = deserialize_events(payload)
    for event in events:
        event_type = _event_type(event)
        if event_type == BLOB_CREATED:
            handler.handle_created(event)
        elif event_type == BLOB_DELETED:
            handler.handle_deleted(event)
        else:
            log.warning("Ignoring unrecognized Event Grid event type: %s", event_type)
    return events


async def receive_events_async(
    payload: JsonPayload,
    handler: AsyncEventHandler,
    logger: logging.Logger | None = None,
) -> list[StructuredEvent]:
    log = logger or logging.getLogger(__name__)
    events = deserialize_events(payload)
    for event in events:
        event_type = _event_type(event)
        if event_type == BLOB_CREATED:
            await handler.handle_created(event)
        elif event_type == BLOB_DELETED:
            await handler.handle_deleted(event)
        else:
            log.warning("Ignoring unrecognized Event Grid event type: %s", event_type)
    return events
